src/dex/hashflow/hashflow.ts

```diff
@@ -0,0 +1,689 @@
+import { HashflowData, PriceLevel, RfqError, RFQType } from './types';
+import { SimpleExchange } from '../simple-exchange';
+import { Adapters, HashflowConfig } from './config';
+import { HashflowApi } from '@hashflow/taker-js';
+import routerAbi from '../../abi/hashflow/HashflowRouter.abi.json';
+import BigNumber from 'bignumber.js';
+import { BN_0, BN_1, getBigNumberPow } from '../../bignumber-constants';
+import { Interface } from 'ethers/lib/utils';
+import { ChainId, ZERO_ADDRESS } from '@hashflow/sdk';
+import {
+  PriceLevelsResponse,
+  RfqResponse,
+} from '@hashflow/taker-js/dist/types/rest';
+import { assert } from 'ts-essentials';
+import {
+  HASHFLOW_BLACKLIST_TTL_S,
+  PRICE_LEVELS_TTL_SECONDS,
+} from './constants';
+
+export class Hashflow extends SimpleExchange implements IDex<HashflowData> {
+  readonly hasConstantPriceLargeAmounts = false;
+  readonly needWrapNative = false;
+  readonly isFeeOnTransferSupported = false;
+  private api: HashflowApi;
+
+  private hashFlowAuthToken: string;
+  private disabledMMs: Set<string>;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(HashflowConfig);
+
+  logger: Logger;
+
+  constructor(
+    readonly network: Network,
+    readonly dexKey: string,
+    readonly dexHelper: IDexHelper,
+    protected adapters = Adapters[network] || {},
+    readonly routerAddress: string = HashflowConfig['Hashflow'][network]
+      .routerAddress,
+    protected routerInterface = new Interface(routerAbi),
+  ) {
+    super(dexHelper, dexKey);
+    this.logger = dexHelper.getLogger(dexKey);
+    const token = dexHelper.config.data.hashFlowAuthToken;
+    assert(
+      token !== undefined,
+      'Hashflow auth token is not specified with env variable',
+    );
+    this.hashFlowAuthToken = token;
+    this.api = new HashflowApi('taker', 'paraswap', this.hashFlowAuthToken);
+    this.disabledMMs = new Set(dexHelper.config.data.hashFlowDisabledMMs);
+  }
+
+  getAdapters(side: SwapSide): { name: string; index: number }[] | null {
+    return this.adapters[side] ? this.adapters[side] : null;
+  }
+
+  getPairName = (srcAddress: Address, destAddress: Address) =>
+    `${srcAddress}_${destAddress}`.toLowerCase();
+
+  getPoolIdentifier(pairName: string, mm: string) {
+    return `${this.dexKey}_${pairName}_${mm}`;
+  }
+
+  async getPoolIdentifiers(
+    srcToken: Token,
+    destToken: Token,
+    side: SwapSide,
+    blockNumber: number,
+  ): Promise<string[]> {
+    const normalizedSrcToken = this.normalizeToken(srcToken);
+    const normalizedDestToken = this.normalizeToken(destToken);
+    const chainId = this.network as ChainId;
+
+    if (normalizedSrcToken.address === normalizedDestToken.address) {
+      return [];
+    }
+
+    const pairName = this.getPairName(
+      normalizedSrcToken.address,
+      normalizedDestToken.address,
+    );
+
+    const cachedLevels = await this.dexHelper.cache.get(
```

Comment:
we can use `cache.getAndCacheLocally`, minor though

Reply:
sry the confusion I mean use both `cache.{setexAndCacheLocally, getAndCacheLocally}` instead of `cache.{set,get}`

Reply:
As we agreed in private, skipping this matter